# Project 3: Build Decision Tree for Bank Loan Modelling

## DV - "Personal Loan"

## IDV - Output of RF Algorithm

In [1]:
import numpy as np
import pandas as pd
from sklearn import tree
from sklearn import preprocessing

In [4]:
df = pd.read_excel("D:/AI-ML/day 25/Bank_Loan.xlsx","Data")
df.head()

ID  Age  Experience  Income  ZIP Code  Family  CCAvg  Education  Mortgage  \
0   1   25           1      49     91107       4    1.6          1         0   
1   2   45          19      34     90089       3    1.5          1         0   
2   3   39          15      11     94720       1    1.0          1         0   
3   4   35           9     100     94112       1    2.7          2         0   
4   5   35           8      45     91330       4    1.0          2         0   

   Personal Loan  Securities Account  CD Account  Online  CreditCard  
0              0                   1           0       0           0  
1              0                   1           0       0           0  
2              0                   0           0       0           0  
3              0                   0           0       0           0  
4              0                   0           0       0           1

In [5]:
df.columns

Index(['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Personal Loan', 'Securities Account',
       'CD Account', 'Online', 'CreditCard'],
      dtype='object')

# Random Forest for selecting important features

In [6]:
from sklearn.ensemble import RandomForestClassifier

In [7]:
df.isna().sum()

ID                    0
Age                   0
Experience            0
Income                0
ZIP Code              0
Family                0
CCAvg                 0
Education             0
Mortgage              0
Personal Loan         0
Securities Account    0
CD Account            0
Online                0
CreditCard            0
dtype: int64

There are no Null values.

In [8]:
objList = df.select_dtypes(include = "object").columns
print (objList)

Index([], dtype='object')


All variables are Numerical.

In [9]:
rf_model=RandomForestClassifier(n_estimators=1000,max_features=2,oob_score=True)

In [10]:
features=['ID', 'Age', 'Experience', 'Income', 'ZIP Code', 'Family', 'CCAvg',
       'Education', 'Mortgage', 'Securities Account',
       'CD Account', 'Online', 'CreditCard']

In [11]:
rf_model.fit(X=df[features],y=df['Personal Loan'])

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features=2,
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=1000,
                       n_jobs=None, oob_score=True, random_state=None,
                       verbose=0, warm_start=False)

In [12]:
print(rf_model.oob_score_)

0.9856


Thus the Model is 98.56% Accurate.

In [13]:
for feature,imp in zip(features,rf_model.feature_importances_):
    print(feature,imp)

ID 0.04371092821652396
Age 0.03866683520357028
Experience 0.038272067177731635
Income 0.3226073119354537
ZIP Code 0.040864198942900984
Family 0.08428790542395714
CCAvg 0.17354605095640488
Education 0.13802249369620026
Mortgage 0.04122621833953404
Securities Account 0.0051550208651919074
CD Account 0.05644182778196517
Online 0.0077684170655216355
CreditCard 0.009430724395044351


## The imp features in this dataset can be: Income,Family,CCAvg, Education.

# Classification

In [14]:
tree_model = tree.DecisionTreeClassifier(max_depth=8)
features=['Income','Family','CCAvg','Education']
tree_model.fit(X=df[features],y=df['Personal Loan'])

DecisionTreeClassifier(ccp_alpha=0.0, class_weight=None, criterion='gini',
                       max_depth=8, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [15]:
with open('loan_Dtree.dot','w') as f:
    f=tree.export_graphviz(tree_model,feature_names=features,out_file=f)

# Inferences from Classification

In [16]:
r=tree.export_text(tree_model,feature_names=features)
print(r)

|--- Income <= 113.50
|   |--- CCAvg <= 2.95
|   |   |--- Income <= 106.50
|   |   |   |--- class: 0
|   |   |--- Income >  106.50
|   |   |   |--- Family <= 3.50
|   |   |   |   |--- Education <= 1.50
|   |   |   |   |   |--- class: 0
|   |   |   |   |--- Education >  1.50
|   |   |   |   |   |--- CCAvg <= 1.65
|   |   |   |   |   |   |--- Family <= 1.50
|   |   |   |   |   |   |   |--- Education <= 2.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Education >  2.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Family >  1.50
|   |   |   |   |   |   |   |--- Income <= 111.50
|   |   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |   |--- Income >  111.50
|   |   |   |   |   |   |   |   |--- class: 1
|   |   |   |   |   |--- CCAvg >  1.65
|   |   |   |   |   |   |--- Education <= 2.50
|   |   |   |   |   |   |   |--- class: 0
|   |   |   |   |   |   |--- Education >  2.50
|   |   |   |   |   |   |   |--- CCAv

# Some of the inferences:

1) People with Income <= 106.50, CCAvg <= 2.95: Personal Loan :"NO"

2)People with 106.50<Income <= 113.50, Family <= 3.50 , Education <= 1.50, CCAvg <= 2.95- Personal Loan: "No"

3)People with Income >  113.50, Education <= 1.50 and Family <= 2.50: Personal Loan: "No"

4) People with Income >  113.50, Education <= 1.50 and Family > 2.50: Personal Loan: "Yes"..etc